### ISSUE TITLE
Optimize bookmarking when "Add duplicates (hash)" is selected

### ISSUE THREAD

**Issue by aberenguel** [sepinf-inc/IPED#1074]
IPED: 4.0.0-snapshot (05de89cf4)

I have a big case (total size of 5TB). When highlight one item and add it to a bookmark (with "Add duplicates (hash)" selected) it takes about 2 minutes with dialog "Searching Duplicates...".

I think it can be optimized, because when I select the item, the IPED already knows which are the duplicated items (as seen in "Duplicated" tab)

![image](https://user-images.githubusercontent.com/2973359/165181099-22d73411-b09f-4e95-a170-2657cb69cf22.png)

**Comment by lfcnassif** [comment]
This should be a performance regression of #179. I'll take a look tomorrow.

**Comment by lfcnassif** [comment]
Just tested this with 3.18.14 and 4.0.0-snapshot on a medium size case (~2M items), it's not a regression, first bookmark creation time is similar on both versions. Second, third... bookmark creation is much faster, have you noticed that? How many items do you have in your case and how much time does it take to create a second bookmark with duplicates for a single item? This first bookmark creation slowness is probably because Lucene should initialize some index structures when they are used at the first time.

The "bookmark duplicates" implementation goal was to make bookmarking of hundreds of thousands of unique items + their duplicates fast. The "Duplicates Tab" implementation is different, since it has to search for just a single unique hash, not hundreds of thousands of different hashes. Maybe the feature could switch between implementations based on the number of hashes being searched for, on case size, if case folder is on SSD or not, not sure if there is a good global decision strategy...

**Comment by aberenguel** [comment]
There are ~44.7M items in the case.

I created a new fresh bookmark with a single item (that it not duplicated) with "Add duplicates (hash)" selected. It still takes a huge time.
The case is RAID HDD (3 disks). The E01 and dd files are on the same partition.

Does the number of bookmarks affect the performance? In this case there are 12 bookmarks with ~400k items already added on them.

**Comment by lfcnassif** [comment]
> Does the number of bookmarks affect the performance?

No, but the number of selected items to bookmark and number of case items yes, since current implementation loops over those lists find duplicates: https://github.com/sepinf-inc/IPED/blob/615eeabf8c515d47a7990dc92e9ab0fcd7868f22/iped-app/src/main/java/dpf/sp/gpinf/indexer/desktop/BookmarksManager.java#L307

**Comment by aberenguel** [comment]
I measured the second loop. It took 113335 ms.

https://github.com/sepinf-inc/IPED/blob/615eeabf8c515d47a7990dc92e9ab0fcd7868f22/iped-app/src/main/java/dpf/sp/gpinf/indexer/desktop/BookmarksManager.java#L341

**Comment by aberenguel** [comment]
I measured the time between the line 343 and the end of the loop https://github.com/sepinf-inc/IPED/blob/615eeabf8c515d47a7990dc92e9ab0fcd7868f22/iped-app/src/main/java/dpf/sp/gpinf/indexer/desktop/BookmarksManager.java#L343

It took an average of 3577 ns

So I commented the `progress.setProgress(++i)`. The average dropped to 27 ns.

I think a solution it to update the progress only when some percentage varies, like 0.1%

**Comment by lfcnassif** [comment]
If you delete the bookmark and create it again, it takes another ~113s?

My test was with 3.18.14/4.0.0-alpha 2M items cases on a SSD disk, first bookmark creation takes about 8-9s to be created, second takes hundreds of milliseconds (I didn't measure precisely), maybe current implementation is badly affected if the case is on HDDs...

**Comment by lfcnassif** [comment]
> So I commented the `progress.setProgress(++i)`. The average dropped to 27 ns.
> 
> I think a solution is to update the progress only when some percentage varies, like 0.1%

W00t great! Updating UI 44M times for sure should be avoided!

**Comment by lfcnassif** [comment]
> I think a solution is to update the progress only when some percentage varies, like 0.1%

Do you have an initial implementation or should I implement it?

**Comment by lfcnassif** [comment]
Just pushed a possible implementation, please check if it works with your huge case.

**Comment by aberenguel** [comment]
Just tested! Now it is taking about 3 seconds.

**Comment by aberenguel** [comment]
Thanks. I wonder if searching by hash would be faster if there are few uniqueSelectedIds.

**Comment by lfcnassif** [comment]
> Just tested! Now it is taking about 3 seconds.

Thanks for debugging the issue and testing!

> Thanks. I wonder if searching by hash would be faster if there are few uniqueSelectedIds.

Possibly yes, just testing to be sure. We can leave this open until this approach is implemented.

**Comment by aberenguel** [comment]
In a few experiments in this case (44M items) I noticed that the time does not depend on how many items was selected. I always take about 2600 ms.

I implemented the hash approach. The difficulty is to decide when use it.
```java
if (includeDuplicatesUsingHash(uniqueSelectedIds.size(), reader.maxDoc())) {
    for (IItemId itemId : uniqueSelectedIds.toArray(new IItemId[uniqueSelectedIds.size()])) {

        IItem item = ipedCase.getItemByItemId(itemId);
        String textQuery = IndexItem.HASH + ":" + item.getHash();
        
        if (progress.isCanceled())
            return;

        IIPEDSearcher searcher = new IPEDSearcher(ipedCase, textQuery);
        for (IItemId dupItem : searcher.multiSearch().getIterator()) {
            uniqueSelectedIds.add(dupItem); // change it -- did not checked if dupItem has already added
        }
    }
    hashed = true;
    return;
}
```

Using the hash approach, the time depends on several factors, like the quantity of items was selected, if it has duplicated items, the ammount of metadata, etc. In some experiments:
- 1 item (no dups) - ~30ms
- 10 items (no dups) - ~170ms
- 100 items (no dups) - ~1100ms
- 200 items (no dups) - ~2000ms
- 400 items (no dups) - ~3900ms

Assuming a linear model, the decisor could be something like:
```java
    private boolean includeDuplicatesUsingHash(int countSelected, int countDocs) {
        return 2600 * countDocs / 44_000_000 > 3900 * countSelected / 400;
    }
```

**Comment by lfcnassif** [comment]
Interesting numbers! Maybe we don't need to always pick the best approach, but switch to the search one if it is the best for sure., using some very low threshold.

And your proposal may be optimized making batch searches for X hashes:
`hash:(hash1 hash2 hash3...)`

**Comment by lfcnassif** [comment]
> I noticed that the time does not depend on how many items was selected

Have you selected the same order of magnitude of the number of case items? Maybe 10/20M items to bookmark could affect the results...

**Comment by aberenguel** [comment]
> Have you selected the same order of magnitude of the number of case items? Maybe 10/20M items to bookmark could affect the results...
I just tried that. Added all items to bookmark. It took the same time (~2600ms).

I'm implementing this feature that uses the hash for few selected items. Soon I push a PR.

**Comment by lfcnassif** [comment]
> I'm implementing this feature that uses the hash for few selected items. Soon I push a PR.

Cool, thank you!

**Comment by aberenguel** [comment]
Changed to use `hash:(hash1 hash2 hash3...)`. New results

1 item (no dups) - ~30ms
10 items (no dups) - ~80ms
100 items (no dups) - ~280ms
200 items (no dups) - ~500ms
400 items (no dups) - ~900ms

**Comment by aberenguel** [comment]
I think there is a regression bug when several items are selected (`includeDuplicatesUsingHash()` is false).

In such scenario, I had ~600k items in table view. So I checked filtering of duplicated resulting in ~140k items. Finally I selected all items and added to a bookmark. Using "Add duplicates (hash)" or not makes no difference. Only the selected items are added. If I selected just few items (so that `includeDuplicatesUsingHash()` is false), it works as expected.
